docs/src/design/many_differentials.md

```diff
@@ -0,0 +1,175 @@
+
+To return to the question of why ChainRules has `Composite{P, <:NamedTuple}` whereas Zygote v0.4 just has `NamedTuple`, it relates to semi-structural derivatives, and being able to overload things more generally.
+If one knows that one has a semi-structural derivative based on property names, like `Composite{QR}(Q=..., R=...)`, and one is adding it to the true structural derivative based on field names `Composite{QR}(factors=..., τ=...)`, then we need to overload the addition operator to perform that correctly.
+We cannot overload happily similar things for `NamedTuple` since we don't know the primal type, only the names of the values contained. 
+In fact we can't actually overload addition at all for `NamedTuple` as that would be type-piracy, so have to use `Zygote.accum` instead.
+
+Another use of the primal being a type parameter is to catch errors. 
+ChainRules disallows the addition of `Composite{SVD}` to `Composite{QR}` since in a correctly differentiated program that can never occur.
+
+### Differentials types for computational efficiency
+
+There is another kind of unnatural differential.
+One that is for computational efficiency.
+ChainRules has `Thunk`s and `InplaceThunk`s, which wrap a computation that computes a derivative and delays that work until it is needed, either via the derivative being added to something or being `unthunk`ed manually.
+Thus saving time if it is never used.
+
+Another differential type used for efficiency is `Zero` which represents the hard zero (in Zygote v0.4 this is `nothing`).
+For example the derivative of `f(x, y)=2x` with respect to `y` is `Zero()`.
+Add `Zero()` to anything, and one gets back the original thing without change.
+We noted that all differentials need to be a vector space.
+ `Zero()` is the [trivial vector space](https://proofwiki.org/wiki/Definition:Trivial_Vector_Space).
+Further, add `Zero()` to any primal value (no matter the type) and you get back another value of the same primal type (the same value in fact).
+So it meets the requirements of a differential type for *all* primal types.
+`Zero` can be a saving on memory since we can avoid allocating anything, and on time since performing the multiplication
+`Zero` and `Thunk` are both examples of a differential type that is valid for multiple primal types.
+
+### Conclusion
+Now, you have seen examples of both differential types that work for multiple primal types, and primal types that have  multiple valid differential types.
+Semantically we can handle these very easily in julia.
+Just put in a few more dispatching on `+`.
+Multiple-dispatch is great like that.
+The down-side is our type-inference becomes hard.
+If you have exactly 1 differential type for each primal type, you can very easily workout what all the types on your reverse pass will be - you don't really need type inference - but you lose so much expressibility.
+
+### Appendix: What Swift does
```

Comment:
This section is kind of weak (and written in first person) because I don't know enough about what Swift is doing